include/bitcoin/bitcoin/math/ring_signature.hpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2011-2018 libbitcoin developers (see AUTHORS)
+ *
+ * This file is part of libbitcoin.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef LIBBITCOIN_RING_SIGNATURE_HPP
+#define LIBBITCOIN_RING_SIGNATURE_HPP
+
+#include <bitcoin/bitcoin/math/elliptic_curve.hpp>
+
+namespace libbitcoin {
+
+/**
+ * key_rings is a collection of public key rings.
+ * Each ring contains several public keys.
+ * To create a valid signature, you must use at least one valid private
+ * key from each ring.
+ * That is given three rings [{A, B, C}, {D, E, F}, {X, Y}], then to
+ * create a valid signature, we must use a private key from each of those
+ * sets.
+ * For example A and E and X. We can summarize this operation as:
+ * (A or B or C) and (D or E or F) and (X or Y)
+ */
+typedef std::vector<point_list> key_rings;
+
+/**
+ * A borromean ring signature.
+ *
+ * theta = {e_0, s_i_j : 0 <= i <= n, 0 <= j <= m_i}
+ */
+struct ring_signature
+{
+    typedef std::vector<secret_list> s_values_type;
```

Comment:
Use existing typedef, avoid "_type" in type names and the suffix "_values" isn't much helpful.
```cpp
struct ring_signature
{
    ec_secret e;
    secret_list s;
};
```

The cryptic single letter variable names are acceptable within the internals of the implementation (only if they are helpful in correlating to some standard), but should be avoided in public interfaces. Assume people don't know what "e" or "s" mean and replace with informative names.